File: pyversion/35. search insert position.py
def searchInsert(nums, target):
    lo = 0
    hi = len(nums) - 1
    while lo <= hi:
        mi = lo + (hi - lo) // 2
        if nums[mi] > target:
            hi = mi - 1
        else:
            lo = mi + 1

    return lo


if __name__ == '__main__':
    nums = [2, 5, 9, 11]
    # print(searchInsert(nums, 1))
    print(searchInsert(nums, 7))
    # print(searchInsert(nums, 15))

File: cppversion/35. search insert position.cpp
#include <iostream>
#include <vector>

int searchInsert(std::vector<int> &nums, int target)
{
    int lo = 0, hi = nums.size() - 1;
    while (lo <= hi)
    {
        int mi = lo + (hi - lo) / 2;
        if (nums[mi] < target)
            lo = mi + 1;
        else
            hi = mi - 1;
    }
    return lo;
}

int main()
{
    std::vector<int> nums = {1, 3, 5, 8, 9};
    std::cout << searchInsert(nums, 10) << std::endl;
    std::cout << searchInsert(nums, 0) << std::endl;
    std::cout << searchInsert(nums, 7) << std::endl;
    std::cin.get();
}
